packages/vite/src/node/server/middlewares/transform.ts

```diff
@@ -74,14 +76,24 @@ export function transformMiddleware(
         if (depsOptimizer?.isOptimizedDepUrl(url)) {
           // If the browser is requesting a source map for an optimized dep, it
           // means that the dependency has already been pre-bundled and loaded
-          const mapFile = url.startsWith(FS_PREFIX)
+          const sourcemapPath = url.startsWith(FS_PREFIX)
             ? fsPathFromId(url)
             : normalizePath(
                 ensureVolumeInPath(path.resolve(root, url.slice(1))),
               )
           try {
-            const map = await loadOptimizedDep(mapFile, depsOptimizer)
-            return send(req, res, map, 'json', {
+            const map = JSON.parse(
+              await loadOptimizedDep(sourcemapPath, depsOptimizer),
+            ) as SourceMap
+
+            applySourcemapIgnoreList(
```

Comment:
Instead of rewriting the sourcemap when serving it, wouldn't it make sense to consider the ignore list predicate when initially generating the optimized dep?

Reply:
I thought about it, but we don't have the .map in memory to modify it. See https://github.com/vitejs/vite/pull/12622, we needed to go back to let esbuild write the files to disk directly because there was a perf regression.
We would also need to start using `sourcemapIgnoreList.toString()` in the hash key for optimization given our current architecture. That could work though.

If we want to do it, I think we should still do it lazily to avoid blocking giving the optimized deps to the browser. But that means that if the process is stopped, we may end up with inconsistent results. 

So I see three options:
a. Current approach in this PR, and we accept the JSON.parse/JSON.stringify cost (I think this could be fine)
b. Current approach, but we don't use JSON.parse/JSON.stringify and instead use regexes to grab the `sources` array and then inject after it the `ignoreSourcemapList`.
c. We add `sourcemapIgnoreList` to the deps cache key, and when we do the first JSON.parse, we also add a marker like `"__vite": "sourcemapIgnorList_done",`. So next time we load it, we can use a regex to know it has been processed.

Leaning towards b. if there is a regression perf after this PR. cc @bluwy @sapphi-red

Reply:
My opinion is to go with "a." for now because I guess this is not a hot path.

One more option would be to create a feature request / PR to esbuild for adding sourcemap ignore list feature.

Reply:
Thanks for the explanation. I'd also heavily lean towards "a." unless there's a serious performance problem.